pandas/tests/arrays/interval/test_interval.py

```diff
@@ -256,9 +256,9 @@ def test_arrow_extension_type():
 
     assert p1.closed == "left"
     assert p1 == p2
-    assert not p1 == p3
+    assert not p1 == p3  # pylint: disable=unneeded-not
     assert hash(p1) == hash(p2)
-    assert not hash(p1) == hash(p3)
+    assert hash(p1) != hash(p3)
```

Comment:
just for my understanding, why does the first one need the check turning off, but the second one can be changed to `LSH != RHS`?

Reply:
I may have been a bit too quick on that one. The first one seems to be a bug either on Pandas or Pyarrow side:
```python
>>> p1 = ArrowIntervalType(pa.int64(), "left")
>>> p3 = ArrowIntervalType(pa.int64(), "right")
>>> p1 == p3
False
>>> p1 != p3
False
>>> p1.equals(p3)
False
```
I looked trough the `ArrowIntervalType` implementation and the `pyarrow.ExtensionType` source and I cannot pinpoint why `p1 != p3` would retourne `False`.

Reply:
🤔 @jorisvandenbossche is this a bug? Does `ArrowIntervalType` need something like
```python
    def __ne__(self, other):
        return not self.__eq__(other)
```
?

Reply:
cc @WillAyd in case you have comments on this

Reply:
Yes, that seems a bug in ArrowIntervalType, and probably needs such a `__ne__` (for example the base ExtensionDtype also has such a method)

Reply:
Actually, checking some docs about this, explicitly implementing `__ne__` should _not_ be needed since Python 3. So not directly sure what is going wrong here ..

Reply:
I'm unsure too. Note that `ArrowPeriodType` has the same behavior.

Reply:
OK, I _think_ what is going on is that the cython class we are inheriting here implements `__eq__` (and therefore maybe also implicitly `__ne__` under the hood?), and then if you subclass that in Python with overriding `__eq__`, it seems that `__ne__` doesn't call the subclass' `__eq__`, but the cython parent class' `__eq__`. 
This might be a gotcha when subclassing cython types? 

From https://stackoverflow.com/questions/4352244/should-ne-be-implemented-as-the-negation-of-eq, they quote:

> The `__ne__` method follows automatically from `__eq__` only if `__ne__` isn't already defined in a superclass. So, if you're inheriting from a builtin, it's best to override both.

It seems that cython defined an `__ne__` for your cython class (even if not doing that explicitly), so as a result when subclassing and overriding `__eq__`, you also need to override `__ne__`

Reply:
(it's probably not cython specific, but just any type that is being defined using the Python C API)

Reply:
Yea very interesting. My guess is that this stems from pyarrow somewhere

```python
>>> getattr(pandas.core.arrays.arrow.extension_types.ArrowIntervalType, "__ne__")
<slot wrapper '__ne__' of 'pyarrow.lib.ExtensionType' objects>
```

If you traverse the mro its interesting to see this change through the inheritance hierarchy:

```python
for cls in pandas.core.arrays.arrow.extension_types.ArrowIntervalType.__mro__: 
    print(cls, getattr(cls, '__ne__'))

<class 'pandas.core.arrays.arrow.extension_types.ArrowIntervalType'> <slot wrapper '__ne__' of 'pyarrow.lib.ExtensionType' objects>
<class 'pyarrow.lib.ExtensionType'> <slot wrapper '__ne__' of 'pyarrow.lib.ExtensionType' objects>
<class 'pyarrow.lib.BaseExtensionType'> <slot wrapper '__ne__' of 'pyarrow.lib.DataType' objects>
<class 'pyarrow.lib.DataType'> <slot wrapper '__ne__' of 'pyarrow.lib.DataType' objects>
<class 'pyarrow.lib._Weakrefable'> <slot wrapper '__ne__' of 'object' objects>
<class 'object'> <slot wrapper '__ne__' of 'object' objects>
```

Reply:
Interestingly enough I didn't see anything in Arrow's build system that would set the language_level for the Cython extension to 3. We do this in setup.py, I think they would ideally do it in their CMake file.

So if its a matter of 2/3 compat with the generated source file could be a nice upstream patch:

https://github.com/apache/arrow/blob/8e3a1e1b7c17213eb8b79fba49c496de6199fd50/cpp/cmake_modules/UseCython.cmake#L138

@lithomas1

Reply:
That said for the scope of this PR I think should just ignore

Reply:
> Interestingly enough I didn't see anything in Arrow's build system that would set the language_level for the Cython extension to 3.

I don't think that matters; if you _run_ Python 3, it has the effect that the default `__ne__` will use `__eq__` automatically. 

But what is happening here is that pyarrow.lib.ExtensionType is a cython class that implements `__eq__`, and cython defaults to implement a `richcompare` slot that also adds `__ne__` based on the `__eq__` of that class. So just like when subclassing a builtin, if you override `__eq__` you also need to override `__ne__` when subclassing a cython class with a `__eq__`.  
This also happens in pandas, for example we have PeriodDtypeBase in cython with `__eq__`, and that initially caused the same issue (https://github.com/pandas-dev/pandas/issues/37265), and that was fixed by adding `__ne__` in the python subclass

Now, we could maybe fix this on the PyArrow side as well by adding an explicit `__ne__` that is defined to be `return not self == other`, which might ensure that subclasses would only need to override `__eq__`, because the parent `__ne__` might then work correctly (something to test). 
(but regardless of that we should fix this on the pandas side)

Reply:
Thanks for your investigations

If this can be resolved as simply as adding
```
    def __ne__(self, other):
        return not self.__eq__(other)
```
(with type annotations), then I'd suggest we do it directly in this PR

The primary motivation for using linters and all it to help catch issues, and I'm glad `pylint` has helped find this one - I'd suggest it just be fixed in this PR